pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go

```diff
@@ -252,3 +253,64 @@ func (d *discoveryNamespacesFilter) isSelected(labels labels.Set) bool {
 
 	return false
 }
+
+type maistraDiscoveryNamespacesFilter struct {
+	lock       sync.RWMutex
+	namespaces sets.String
+}
+
+func NewMaistraDiscoveryNamespacesFilter(mrc memberroll.MemberRollController) DiscoveryNamespacesFilter {
+	d := &maistraDiscoveryNamespacesFilter{}
+	mrc.Register(d, "MaistraDiscoveryNamespacesFilter")
+	return d
+}
+
+func (d *maistraDiscoveryNamespacesFilter) SetNamespaces(namespaces ...string) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.namespaces = sets.NewString(namespaces...)
+}
+
+func (d *maistraDiscoveryNamespacesFilter) Filter(obj interface{}) bool {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	return d.namespaces.Has(obj.(metav1.Object).GetNamespace())
+}
+
+// SelectorsChanged is not implemented, because Maistra does not filter namespaces by selectors;
+// this function should never be called if memberRollName is specified, so it's safe to call panic.
+func (d *maistraDiscoveryNamespacesFilter) SelectorsChanged(_ []*metav1.LabelSelector) (selectedNamespaces []string, deselectedNamespaces []string) {
+	panic("Not implemented")
+}
+
+// SyncNamespaces does not need to do anything, because namespaces are set by the serviceMeshMemberRollListener
+// that reacts on ServiceMeshMemberRoll changes.
+func (d *maistraDiscoveryNamespacesFilter) SyncNamespaces() error {
+	return nil
+}
```

Comment:
@luksa I had to implement this function, because it was added to the interface in the meantime. Could you take a look a this? I added some comments, because I had to investigate why some functions call `panic` and why are not needed.

Reply:
Strange- this shouldn't be necessary. The function should only be called when `nsLister != nil`

Reply:
Ah right... I forgot that when `maistraDiscoveryNamespacesFilter`, then `nsLister` is not used.

Reply:
But returning nil is also correct and I think it's safer, because I can imagine that the upstream removes this condition, we merge it without conflict and then our Istio is broken.